app/routes.py

```diff
@@ -1 +1,268 @@
+###~~~--- task model routes ---~~~###
+# - - - - - - - - - - - - - - - - - #
+
+task_bp = Blueprint("tasks", __name__, url_prefix="/tasks")
+
+@task_bp.route("", methods=['POST'])
+
+# define a route for creating a task
+def create_task():
+    request_body = request.get_json()
+
+    try:
+        new_task = Task(
+            title = request_body["title"],
+            description = request_body["description"],
+        )
+    except:
+        abort(make_response({"details": f"Invalid data"}, 400))
+
+    db.session.add(new_task)
+    db.session.commit()
+
+    return {"task": new_task.to_dict()}, 201
+
+# define a route for getting all tasks
+@task_bp.route("", methods=["GET"])
+def read_all_tasks():
+    # querries
+    title_query = request.args.get("title")
+    description_query = request.args.get("description")
+    completed_at_query = request.args.get("completed_at")
+    sort_query = request.args.get("sort")
+
+    if title_query:
+        tasks = Task.query.filter_by(title=title_query)
+    
+    elif description_query:
+        tasks = Task.query.filter_by(description=description_query)
+    
+    elif completed_at_query:
+        tasks = Task.query.filter_by(completed_at=completed_at_query)
+    
+    elif sort_query:
+        if sort_query == 'asc':
+            tasks = Task.query.order_by(Task.title).all()
+        elif sort_query == 'desc':
+            tasks = Task.query.order_by(Task.title.desc()).all()
+    
+    else:
+        tasks = Task.query.all()
+
+    tasks_response = []
+
+    for task in tasks:
+        tasks_response.append(task.to_dict())
+
+    return jsonify(tasks_response), 200
+
+@task_bp.route("/<task_id>", methods=["GET"])
+def read_one_task(task_id):
+    task = validate_model(Task, task_id)
+
+    if task.goal_id:
+        return {
+        "task": {
+            "id": task.task_id,
+            "goal_id": task.goal_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete": True if task.completed_at else False
+        }
+    }, 200
+
+    return {"task": task.to_dict()}, 200
+
+@task_bp.route("/<task_id>", methods=["PUT"])
+def update_task(task_id):
+    task = validate_model(Task, task_id)
+
+    request_body = request.get_json()
+
+    task.title = request_body["title"]
+    task.description = request_body["description"]
+
+    db.session.commit()
+
+    return {"task": task.to_dict()}, 200
+
+
+@task_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def task_mark_complete(task_id):
+    task = validate_model(Task, task_id) 
+
+    task.completed_at = date.today()
+
+    db.session.commit()
+
+    key = os.environ.get("key")
+    url_path = "https://slack.com/api/chat.postMessage"
+    body = {
+        "channel": "C056W45DVEH",
+        "text": f"Someone just completed the task {task.title}"
+    }
+
+    header = {"Authorization": f"Bearer {key}"}
+
+    slack = requests.post(url_path, headers=header, json=body)
```

Comment:
Nice work on having your Slack post sent after the logic of marking a task complete! We don't want to send out any false positive alerts just in case our logic fails during the update!